go-controller/pkg/clustermanager/pod/ip_allocator.go

```diff
@@ -0,0 +1,265 @@
+	releaseIPsFromAllocator := false
+
+	for _, obj := range objs {
+		pod, ok := obj.(*corev1.Pod)
+		if !ok {
+			klog.Errorf("Could not cast %T object to *corev1.Pod", obj)
+			continue
+		}
+		err := a.reconcile(nil, pod, releaseIPsFromAllocator)
+		if err != nil {
+			klog.Errorf("Failed to sync pod %s/%s: %w", pod.Namespace, pod.Name, err)
+		}
+	}
+
+	return nil
+}
+
+func (a *PodIPAllocator) reconcile(old, new *corev1.Pod, releaseIPsFromAllocator bool) error {
+	var pod *corev1.Pod
+	if old != nil {
+		pod = old
+	}
+	if new != nil {
+		pod = new
+	}
+
+	podScheduled := util.PodScheduled(pod)
+	podWantsHostNetwork := util.PodWantsHostNetwork(pod)
+
+	// nothing to do for a unscheduled or host network pods
+	if !podScheduled || podWantsHostNetwork {
+		return nil
+	}
+
+	onNetwork, networkMap, err := util.GetPodNADToNetworkMapping(pod, a.netInfo)
+	if err != nil {
+		return fmt.Errorf("failed to get NAD to network mapping: %w", err)
+	}
+
+	// nothing to do if not on this network
+	// Note: we are not considering a hotplug scenario where we would have to
+	// release IPs if the pod was unplugged from the network
+	if !onNetwork {
+		return nil
+	}
+
+	// reconcile for each NAD
+	for nadName, network := range networkMap {
+		err = a.reconcileForNAD(old, new, nadName, network, releaseIPsFromAllocator)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (a *PodIPAllocator) reconcileForNAD(old, new *corev1.Pod, nad string, network *nettypes.NetworkSelectionElement, releaseIPsFromAllocator bool) error {
+	var pod *corev1.Pod
+	if old != nil {
+		pod = old
+	}
+	if new != nil {
+		pod = new
+	}
+	podDeleted := new == nil
+	podCompleted := util.PodCompleted(pod)
+
+	if podCompleted || podDeleted {
+		return a.releasePodOnNAD(pod, nad, podDeleted, releaseIPsFromAllocator)
+	}
+
+	return a.allocatePodOnNAD(pod, nad, network)
+}
+
+func (a *PodIPAllocator) releasePodOnNAD(pod *corev1.Pod, nad string, podDeleted, releaseIPsFromAllocator bool) error {
+	if !util.DoesNetworkRequireIPAM(a.netInfo) {
+		// no need to release if no IPAM
+		return nil
+	}
+
+	podAnnotation, _ := util.UnmarshalPodAnnotation(pod.Annotations, nad)
+	if podAnnotation == nil {
+		// track release pods even if they have no annotation in case a user
+		// might have removed it manually
+		podAnnotation = &util.PodAnnotation{}
+	}
+
+	uid := string(pod.UID)
+
+	// do not release IPs from the allocator if not flaged to do so or if they
+	// werea already previosuly released
+	if releaseIPsFromAllocator && !a.isPodReleased(nad, uid) {
+		err := a.allocator.ReleaseIPs(a.netInfo.GetNetworkName(), podAnnotation.IPs)
+		if err != nil {
+			return fmt.Errorf("failed to release ips %v for pod %s/%s and nad %s: %w",
+				util.StringSlice(podAnnotation.IPs),
+				pod.Name,
+				pod.Namespace,
+				nad,
+				err,
+			)
+		}
+	}
+
+	if podDeleted {
+		a.deleteReleasedPod(nad, string(pod.UID))
+	} else {
+		a.addReleasedPod(nad, string(pod.UID))
+	}
+
+	return nil
```

Comment:
nit

```suggestion
	} 
	return a.addReleasedPod(nad, string(pod.UID))
```

Reply:
Can't do. `addReleasedPod`  does not return anything.

Reply:
ah! my turn to facepalm :)